### PR TITLE
[BE] explicitly export subpackage `torch.utils`

### LINKED ISSUE #126401: TYP: declare torch.utils in __all__
### 🐛 Describe the bug

`torch.utils` is accessible but utils is not declared in `__all__` in `torch/.__init__.py` so type checkers do not know it exists.

```py
impor torch

torch.utils.data.Dataset  # pyright: "utils" is not a known attribute of module "torch" (reportAttributeAccessIssue)
```

### Versions

Collecting environment information...
PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:51:49) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] onnx==1.16.0
[pip3] tf2onnx==1.16.1
[pip3] torch==2.3.0
[pip3] torchaudio==2.3.0
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torchaudio                2.3.0                    pypi_0    pypi

cc @ezyang @malfet @xuzhao9 @gramster @albanD

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127705
* #127704
* #127706
* __->__ #128342
* #127707

Resolves #126401

cc @albanD

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/utils/__init__.py [hunk 1]
```diff
@@ -1,14 +1,23 @@
 # mypy: allow-untyped-defs
-import os.path as _osp
-import torch
 
-from .throughput_benchmark import ThroughputBenchmark
-from .cpp_backtrace import get_cpp_backtrace
-from .backend_registration import rename_privateuse1_backend, generate_methods_for_privateuse1_backend
-from . import deterministic
-from . import collect_env
```
Comment (Skylion007) [inline torch/utils/__init__.py:9]: If we want to ban relative imports, there is a ruff rule to do that.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1,20 +1,33 @@
 # mypy: allow-untyped-defs
-from collections import OrderedDict, namedtuple
+
+import functools
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torch/nn/modules/module.py:3]: I will do it in a follow-up.

File: torch/_tensor.py [hunk 3]
```diff
@@ -10,7 +10,6 @@
 
 import torch
 import torch._C as _C
-import torch.utils.hooks as hooks
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torch/_tensor.py:13]: I export `torch.utils.data` to `torch.utils` while `torch.utils.data` requires `torch.Tensor`.

File: torch/utils/__init__.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/utils/__init__.py:9]: I suggested that before in https://github.com/pytorch/pytorch/pull/127703#discussion_r1623617154. Relative imports are useful if the subpackage hierarchy is very deep. I'm open to that.

**Pull request closed without merging**